utils/src/expiring_cache.rs

```diff
@@ -0,0 +1,152 @@
+use arc_swap::ArcSwapOption;
+use std::{
+    future::Future,
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+    time::{Duration, Instant},
+};
+
+struct Entry<T> {
+    item: T,
+    timestamp: Instant,
+}
+
+/// An expiring cache for a single object
+pub struct ExpiringCache<T> {
+    store: ArcSwapOption<Entry<T>>,
+    refetch: Duration,
+    expire: Duration,
+    fetching: AtomicBool,
+}
+
+impl<T: Clone> ExpiringCache<T> {
+    /// Constructs a new expiring cache where `fetch` is the amount of time required to trigger a data
+    /// refetch and `expire` is the time duration after which the stored item is guaranteed not to be returned.
+    ///
+    /// Panics if `refetch > expire`.
+    pub fn new(refetch: Duration, expire: Duration) -> Self {
+        assert!(refetch <= expire);
+        Self { store: Default::default(), refetch, expire, fetching: Default::default() }
+    }
+
+    /// Returns the cached item or possibly fetches a new one using the `refetch_future` task. The
+    /// decision whether to refetch depends on the configured expiration and refetch times for this cache.  
+    pub async fn get<F>(&self, refetch_future: F) -> T
+    where
+        F: Future<Output = T> + Send + 'static,
+        F::Output: Send + 'static,
+    {
+        let mut fetching = false;
+
+        {
+            let guard = self.store.load();
+            if let Some(entry) = guard.as_ref() {
+                if let Some(elapsed) = Instant::now().checked_duration_since(entry.timestamp) {
+                    if elapsed < self.refetch {
+                        return entry.item.clone();
+                    }
+                    // Refetch is triggered, attempt to capture the task
+                    fetching = self.fetching.compare_exchange(false, true, Ordering::SeqCst, Ordering::SeqCst).is_ok();
+                    // If the fetch task is not captured and expire time is not over yet, return with prev value. Another
+                    // thread is refetching the data but we can return with the not-too-old value
+                    if !fetching && elapsed < self.expire {
+                        return entry.item.clone();
+                    }
+                }
+                // else -- In rare cases where now < timestamp, fall through to re-update the cache
+            }
+        }
+
+        // We reach here if either we are the refetching thread or the current data has fully expired
+        let new_item = refetch_future.await;
+        let timestamp = Instant::now();
+        // Update the store even if we were not in charge of refetching - let the last thread make the final update
+        self.store.store(Some(Arc::new(Entry { item: new_item.clone(), timestamp })));
+
+        if fetching {
+            let result = self.fetching.compare_exchange(true, false, Ordering::SeqCst, Ordering::SeqCst);
```

Comment:
why not `.store(false, Release)`

Reply:
I personally like the compare exchange bcs it allows the assert below, and is the more "accurate" operation here